lxd/metrics/api_rates.go

```diff
@@ -0,0 +1,167 @@
+package metrics
+
+import (
+	"net/url"
+	"strings"
+	"sync/atomic"
+
+	"github.com/canonical/lxd/shared/entity"
+	"github.com/canonical/lxd/shared/version"
+)
+
+// requestResult represents a completed request status category.
+type requestResult string
+
+const (
+	errorServer requestResult = "error_server"
+	errorClient requestResult = "error_client"
+	success     requestResult = "succeeded"
+)
+
+// PossibleRequestResults contains all the possible request result categories.
+var PossibleRequestResults = []requestResult{
+	errorServer,
+	errorClient,
+	success,
+}
+
+// resolveReqResult categorizes a status code between serverError, clientError and success.
+// The status code could be coming from an asynchronous operation or a request response.
+func resolveReqResult(statusCode int, asyncOperation bool) requestResult {
+	// Status codes from operations follow a different pattern.
+	if asyncOperation {
+		// If status is not 200(succeeded) nor 401(cancelled), it is considered a failure.
+		if statusCode != 200 && statusCode != 401 {
+			return errorServer
+		}
+
+		return success
+	}
+
+	// 4* codes are considered client errors on HTTP.
+	if statusCode >= 400 && statusCode < 500 {
+		return errorClient
+	}
+
+	// 2* codes are considered successes on HTTP, 300 are for forwarded requests
+	// that are included as successes for the purpose of the metrics.
+	if statusCode >= 200 && statusCode < 400 {
+		return success
+	}
+
+	return errorServer
+}
+
+// resolveEntityType derives the main entity type an endpoint is opertaing on from its URL.
+// This is used to categorize completed requests in internal metrics.
+func resolveEntityType(url url.URL) entity.Type {
+	path := url.Path
+	if url.RawPath != "" {
+		path = url.RawPath
+	}
+
+	// For /, /{version} and /internal endpoints.
+	if !strings.HasPrefix(path, "/"+version.APIVersion+"/") {
```

Comment:
As the default is type.Server at the bottom we dont need this bit as its redundant.

Reply:
Suggest you change the logic on line 68 to handle URLs without a top-level entity type, and then rather than switching on pathParts[0] you switch on something more readable like "urlPrefix" and it can be empty for the server level URLs so that the default case catches it.